django/db/models/query.py

```diff
@@ -889,25 +890,50 @@ def bulk_update(self, objs, fields, batch_size=None):
         requires_casting = connection.features.requires_casted_case_in_updates
         batches = (objs[i : i + batch_size] for i in range(0, len(objs), batch_size))
         updates = []
+        resolvable_fields = set()
         for batch_objs in batches:
             update_kwargs = {}
             for field in fields:
                 when_statements = []
                 for obj in batch_objs:
                     attr = getattr(obj, field.attname)
-                    if not hasattr(attr, "resolve_expression"):
+                    if hasattr(attr, "resolve_expression"):
+                        resolvable_fields.add(field.attname)
+                    else:
                         attr = Value(attr, output_field=field)
                     when_statements.append(When(pk=obj.pk, then=attr))
                 case_statement = Case(*when_statements, output_field=field)
                 if requires_casting:
                     case_statement = Cast(case_statement, output_field=field)
                 update_kwargs[field.attname] = case_statement
-            updates.append(([obj.pk for obj in batch_objs], update_kwargs))
+            updates.append(({obj.pk: obj for obj in batch_objs}, update_kwargs))
         rows_updated = 0
         queryset = self.using(self.db)
         with transaction.atomic(using=self.db, savepoint=False):
-            for pks, update_kwargs in updates:
-                rows_updated += queryset.filter(pk__in=pks).update(**update_kwargs)
+            if resolvable_fields and connection.features.can_return_rows_from_update:
```

Comment:
Should we take the same approach as `Model.save` to

1. Return the dependent generated fields
2. Clear the cached values of fields meant to be returned (generated, expressions) when `not can_return_rows_from_update`